cpp/src/arrow/util/decimal_test.cc

```diff
@@ -590,13 +597,62 @@ TEST(TestDecimalFromRealFloat, LargeValues) {
   // Test the entire float range
   for (int32_t scale = -38; scale <= 38; ++scale) {
     float real = std::pow(10.0f, static_cast<float>(scale));
-    CheckDecimalFromRealIntegerString(real, 1, -scale, "1");
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 1, -scale, "1");
+  }
+  for (int32_t scale = -37; scale <= 36; ++scale) {
+    float real = 123.f * std::pow(10.f, static_cast<float>(scale));
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 2, -scale - 1, "12");
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 3, -scale, "123");
+    CheckDecimalFromRealIntegerString<Decimal128>(real, 4, -scale + 1, "1230");
+  }
+}
+
+// Same tests, repeated for Decimal256
+class TestDecimal256FromRealFloat : public ::testing::TestWithParam<FromFloatTestParam> {
+};
+
+TEST_P(TestDecimal256FromRealFloat, SuccessConversion) {
+  const auto param = GetParam();
+  CheckDecimalFromReal<Decimal256>(param.real, param.precision, param.scale,
+                                   param.expected);
+}
+
+// clang-format off
+INSTANTIATE_TEST_SUITE_P(
+    TestDecimal256FromRealFloat, TestDecimal256FromRealFloat,
+    ::testing::Values(
+        // 2**63 + 2**40 (exactly representable in a float's 24 bits of precision)
+        FromFloatTestParam{9.223373e+18f, 19, 0, "9223373136366403584"},
+        FromFloatTestParam{-9.223373e+18f, 19, 0, "-9223373136366403584"},
+        FromFloatTestParam{9.223373e+14f, 19, 4, "922337313636640.3584"},
+        FromFloatTestParam{-9.223373e+14f, 19, 4, "-922337313636640.3584"},
+        // 2**64 - 2**40 (exactly representable in a float)
+        FromFloatTestParam{1.8446743e+19f, 20, 0, "18446742974197923840"},
+        FromFloatTestParam{-1.8446743e+19f, 20, 0, "-18446742974197923840"},
+        // 2**64 + 2**41 (exactly representable in a float)
+        FromFloatTestParam{1.8446746e+19f, 20, 0, "18446746272732807168"},
+        FromFloatTestParam{-1.8446746e+19f, 20, 0, "-18446746272732807168"},
+        FromFloatTestParam{1.8446746e+15f, 20, 4, "1844674627273280.7168"},
+        FromFloatTestParam{-1.8446746e+15f, 20, 4, "-1844674627273280.7168"},
+        // Almost 10**38 (minus 2**103)
+        FromFloatTestParam{9.999999e+37f, 38, 0,
+                           "99999986661652122824821048795547566080"},
+        FromFloatTestParam{-9.999999e+37f, 38, 0,
+                           "-99999986661652122824821048795547566080"}
+));
+// clang-format on
+
+TEST(TestDecimal256FromRealFloat, LargeValues) {
```

Comment:
Perhaps fold this test into `TestDecimalFromReal`, to avoid repeating it for both decimal types?